src/ledger/recovery.zig

```diff
@@ -0,0 +1,500 @@
+    defer organized.deinit();
+    errdefer organized.deinitShreds();
+
+    // Reconstruct the shard bytes using reed solomon
+    var reed_solomon = try reed_solomon_cache.get(
+        @intCast(meta.code_header.num_data_shreds),
+        @intCast(meta.code_header.num_code_shreds),
+    );
+    defer reed_solomon.deinit();
+    try reed_solomon.reconstruct(allocator, organized.shards, false);
+
+    // Reconstruct code and data shreds from erasure encoded shards.
+    const all_shreds = try reconstructShreds(allocator, meta, organized);
+    defer allocator.free(all_shreds);
+
+    // Assemble list that excludes the shreds we already had
+    var ret = try std.ArrayList(Shred).initCapacity(allocator, organized.num_to_recover);
+    for (all_shreds, organized.mask) |shred, was_present| {
+        if (!was_present) {
+            try shred.sanitize();
+            ret.appendAssumeCapacity(shred);
+        }
+    }
+    return ret;
+}
+
+/// Collect metadata from the shreds about their erasure set,
+/// to be used during shred recovery.
+///
+/// Input shreds are expected to all be from the same erasure set,
+/// and there should be at least one coding shred.
+///
+/// Grab {common, code} headers from first code shred.
+/// Incoming shreds are resigned immediately after signature verification,
+/// so we can just grab the retransmitter signature from one of the
+/// available shreds and attach it to the recovered shreds.
+fn getRecoveryMetadata(shreds: []const Shred) !RecoveryMetadata {
+    const meta: RecoveryMetadata = for (shreds) |shred| {
+        if (shred == .code) {
+            const code_shred = shred.code;
+            const chained_merkle_root = code_shred.fields.chainedMerkleRoot() catch null;
+            const retransmitter_signature = code_shred.fields.retransmitterSignature() catch null;
+            const position = code_shred.fields.custom.position;
+            var common_header = code_shred.fields.common;
+            var code_header = code_shred.fields.custom;
+            common_header.index = try checkedSub(common_header.index, position);
+            code_header.position = 0;
+            break .{
+                .common_header = common_header,
+                .code_header = code_header,
+                .chained_merkle_root = chained_merkle_root,
+                .retransmitter_signature = retransmitter_signature,
+            };
+        }
+    } else return error.TooFewParityShards;
+    std.debug.assert(verifyErasureBatch(meta.common_header, meta.code_header, shreds));
+    std.debug.assert(!meta.common_header.variant.resigned or
+        meta.retransmitter_signature != null);
+    return meta;
+}
+
+fn organizeShredsForRecovery(
+    allocator: Allocator,
+    shreds: []const Shred,
+    meta: RecoveryMetadata,
+) !RecoveryShreds {
+    const num_shards: usize =
+        @intCast(meta.code_header.num_data_shreds + meta.code_header.num_code_shreds);
```

Comment:
Don't think you need this `@intCast`, `num_data_shreds` and `num_code_shreds` are `u16`s.

Reply:
9d24462d596fb2aae458c3483045e2240155f468